ibis/backends/base/sql/registry/window.py

```diff
@@ -42,201 +27,140 @@
-        assert f is None or f >= 0
-
-        if f is None:
-            prefix = 'UNBOUNDED'
+        if frame.end is None:
+            end = 'UNBOUNDED FOLLOWING'
         else:
-            if not f:
-                return 'CURRENT ROW'
-            prefix = str(f)
-
-        return f'{prefix} FOLLOWING'
-
-    if translator._forbids_frame_clause and isinstance(
-        op.expr, translator._forbids_frame_clause
-    ):
-        frame = None
-    elif p is not None and f is not None:
-        frame = f'{window.how.upper()} BETWEEN {_prec(p)} AND {_foll(f)}'
-    elif p is not None:
-        if isinstance(p, tuple):
-            start, end = p
-            frame = '{} BETWEEN {} AND {}'.format(
-                window.how.upper(), _prec(start), _prec(end)
-            )
-        else:
-            kind = 'ROWS' if p > 0 else 'RANGE'
-            frame = f'{kind} BETWEEN {_prec(p)} AND UNBOUNDED FOLLOWING'
-    elif f is not None:
-        if isinstance(f, tuple):
-            start, end = f
-            frame = '{} BETWEEN {} AND {}'.format(
-                window.how.upper(), _foll(start), _foll(end)
-            )
-        else:
-            kind = 'ROWS' if f > 0 else 'RANGE'
-            frame = f'{kind} BETWEEN UNBOUNDED PRECEDING AND {_foll(f)}'
-    else:
-        frame = None
+            end = format_window_boundary(translator, frame.end)
 
-    if frame is not None:
+        frame = f'{frame.how.upper()} BETWEEN {start} AND {end}'
         components.append(frame)
 
     return 'OVER ({})'.format(' '.join(components))
 
 
-_subtract_one = '({} - 1)'.format
-
-
-_expr_transforms = {
-    ops.RowNumber: _subtract_one,
-    ops.DenseRank: _subtract_one,
-    ops.MinRank: _subtract_one,
-    ops.NTile: _subtract_one,
-}
-
-
 def window(translator, op):
-    arg, window = op.args
-
     _unsupported_reductions = (
         ops.ApproxMedian,
         ops.GroupConcat,
         ops.ApproxCountDistinct,
     )
 
-    if isinstance(arg, _unsupported_reductions):
+    if isinstance(op.func, _unsupported_reductions):
         raise com.UnsupportedOperationError(
-            f'{type(arg)} is not supported in window functions'
+            f'{type(op.func)} is not supported in window functions'
         )
 
-    if isinstance(arg, ops.CumulativeOp):
-        arg = cumulative_to_window(translator, arg, window)
+    if isinstance(op.func, ops.CumulativeOp):
+        arg = cumulative_to_window(translator, op.func, op.frame)
         return translator.translate(arg)
 
     # Some analytic functions need to have the expression of interest in
     # the ORDER BY part of the window clause
-    if isinstance(arg, translator._require_order_by) and not window._order_by:
-        window = window.order_by(arg.args[0])
+    frame = op.frame
+    if isinstance(op.func, translator._require_order_by) and not frame.order_by:
+        frame = frame.copy(order_by=(op.func.arg,))
 
     # Time ranges need to be converted to microseconds.
```

Comment:
Is this comment still accurate? If you haven't already done it in this PR we should remove the microseconds assumption here and try to move that logic into the specific backend that needs it (I believe it's BigQuery?).

Reply:
Yes, it is the bigquery backend. I'd rather do that in the follow-up, since I cannot run the impala tests locally which uses the same base backend.

Reply:
Created follow-up https://github.com/ibis-project/ibis/issues/5419